### PR TITLE
feat(faceted-search):  bring the faceted search to ui

### PR DESCRIPTION
**Pull request by nicomaligne**
**What is the problem this PR is trying to solve?**
The first iteration of the faceted search has been developed on data inventory.
The component is still due to changing in the near future and it is still considered in progress.
[UX](http://guidelines.talend.com/document/215501#/navigation-layout/filtering-search)
**What is the chosen solution to this problem?**
Bringing the component to UI will allow other projects to use it and contribute to future evolution.
There are still things missing in this pr, but it is already quite huge.
What will come next :
- A solid documentation  ([WIP](https://github.com/Talend/ui-dataset/pull/40))
-  More storybook example
- More tests
- New badges
- Creating badge from url
- ...

**Please check if the PR fulfills these requirements**

* [ ] The PR commit message follows our [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md)
* [ ] Tests for the changes have been added (for bug fixes / features) And [non reg](./screenshots.md) done before need review
* [ ] Docs have been added / updated (for bug fixes / features)
* [ ] Related design / discussions / pages (not in jira), if any, are all linked or available in the PR

**[ ] This PR introduces a breaking change**

### CODE REVIEW

File: packages/faceted-search/src/components/AddFacetPopover/AddFacetPopover.component.js [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+};
+
+AddFacetRow.propTypes = {
+	id: PropTypes.string.isRequired,
+	badgeDefinition: badgeFacetedPropTypes.isRequired,
+	label: PropTypes.string.isRequired,
+	onClick: PropTypes.func.isRequired,
+};
+
+const filterByAttribute = attribute => badgeDefinition =>
```
Comment (vbocquet) [inline packages/faceted-search/src/components/AddFacetPopover/AddFacetPopover.component.js]: I think you need to trim/downcase filter attribute: If a search `Name`, it didn't filter `Name` badge. On the other side, `name` find `Name`. Same for `    name` which doesn't works.

File: packages/faceted-search/src/components/FacetedToolbar/FacetedToolbar.component.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+import PropTypes from 'prop-types';
+import Action from '@talend/react-components/lib/Actions/Action';
+import { getTheme } from '@talend/react-components/lib/theme';
+
+import { FACETED_MODE } from '../../constants';
+import cssModule from './FacetedToolbar.scss';
+
+const theme = getTheme(cssModule);
+
+const SwitchFacetedMode = ({ facetedMode, id, onClickAdvanced, onClickClassic, t }) => (
```
Comment (vbocquet) [inline packages/faceted-search/src/components/FacetedToolbar/FacetedToolbar.component.js]: I don't remember what you say before Kickoff, but we are ok that, for now, switch mode from basic to advanced do not convert badge to TQL query?

File: packages/faceted-search/src/components/FacetedToolbar/FacetedToolbar.scss [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+				height: inherit;
+			}
+			&:active,
+			&:focus,
+			&:hover {
+				background: $st-tropaz;
+			}
+		}
+
+		&-advanced {
```
Comment (vbocquet) [inline packages/faceted-search/src/components/FacetedToolbar/FacetedToolbar.scss]:
```suggestion
		&-advanced, &-advanced.tc-button-selected {
```
There is a visual bug when advanced action is selected: radius is add to action, here is proposition of fix

File: packages/faceted-search/src/components/FacetedToolbar/FacetedToolbar.scss [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+			border-radius: $padding-large 0 0 $padding-large;
+		}
+		&-basic {
```
Comment (vbocquet) [inline packages/faceted-search/src/components/FacetedToolbar/FacetedToolbar.scss]:
```suggestion
		&-basic, &-basic.tc-button-selected {
```
Same here

File: packages/faceted-search/src/components/facetedSearch.propTypes.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
+import PropTypes from 'prop-types';
+
+const badgeDefinitionRawDataPropTypes = PropTypes.shape({
+	attribute: PropTypes.string.isRequired,
+	label: PropTypes.string.isRequired,
+	operators: PropTypes.arrayOf(PropTypes.string),
+	type: PropTypes.string.isRequired,
+	metadata: PropTypes.shape({
+		badges_per_facet: PropTypes.string,
```
Comment (vbocquet) [inline packages/faceted-search/src/components/facetedSearch.propTypes.js]: What is this key?

File: packages/faceted-search/src/components/Badges/BadgeFaceted/BadgeFaceted.component.js [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+	const onSubmitBadge = event => {
+		overlayDispatch(OVERLAY_FLOW_ACTIONS.closeAll);
+		event.preventDefault();
+		dispatch(
+			BADGES_ACTIONS.update(badgeId, {
+				value: badgeValue,
+				operator: badgeOperator,
+			}),
+		);
+		onSubmit(event, state.badges);
```
Comment (vbocquet) [inline packages/faceted-search/src/components/Badges/BadgeFaceted/BadgeFaceted.component.js]: Think i find a bug, here is the pattern:
- Start storybook, story FacetedSearch > default
- Add a text badge with value `TEST1` and apply
- On Actions log, you see a `onSubmit` event, but there is no `value` key on `properties`
- Click to edit to badge and change value to `TEST2`, then apply
- On Actions log, you see a `onSubmit` event, but there is no `value` key on `properties`

In my opinion, you use `state.badges` as args here, but is not already "muted" with your value. You need to compute an object here to be sure that your arg state is the good one

File: packages/faceted-search/src/components/FacetedManager/FacetedManager.component.js [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+		getOperatorsFromDict,
+		id,
+		inProgress,
+		t,
+	};
+	return <FacetedSearchProvider value={contextValue}>{children}</FacetedSearchProvider>;
+};
+
+FacetedManager.propTypes = {
+	children: PropTypes.oneOfType(PropTypes.element, PropTypes.arrayOf(PropTypes.element)),
```
Comment (vbocquet) [inline packages/faceted-search/src/components/FacetedManager/FacetedManager.component.js]:
```suggestion
	children: PropTypes.oneOfType([PropTypes.element, PropTypes.arrayOf(PropTypes.element)]),
```

File: packages/faceted-search/src/components/AddFacetPopover/AddFacetPopover.component.js [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+					placeholder={t('ADD_FACET_FILTER_PLACEHOLDER', {
+						defaultValue: 'Find a filter',
+					})}
+					onToggle={resetFilter}
+					onFilter={onFilter}
+					value={filterValue}
+				/>
+				{badgesDefinitionsFaceted.map(badgeDefinition => (
+					<AddFacetRow
+						badgeDefinition={badgeDefinition}
```
Comment (vbocquet) [inline packages/faceted-search/src/components/AddFacetPopover/AddFacetPopover.component.js]: Need a `key` here

File: packages/faceted-search/src/queryClient/index.js [hunk 9]
```diff
@@ -0,0 +1,4 @@
+import { createTqlQuery } from './tql';
+
+// eslint-disable-next-line import/prefer-default-export
+export { createTqlQuery };
```
Comment (vbocquet) [inline packages/faceted-search/src/queryClient/index.js]: This is what we import in our app to convert badges to tql?

**Comment by vbocquet** [comment]
Just another global comment about default export. Maybe we can use it more in your file to prevent usage of `eslint-disable`

File: packages/faceted-search/src/components/AdvancedSearch/AdvancedSearch.component.js [hunk 10]
```diff
@@ -0,0 +1,137 @@
+import React, { useState } from 'react';
+import PropTypes from 'prop-types';
+import keycode from 'keycode';
+import { Icon, Action, CircularProgress } from '@talend/react-components';
```
Comment (jmfrancois) [inline packages/faceted-search/src/components/AdvancedSearch/AdvancedSearch.component.js]: please use direct import (using /lib/Icon)

File: packages/faceted-search/src/components/FacetedSearchIcon/FacetedSearchIcon.component.js [hunk 11]
```diff
@@ -0,0 +1,31 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { ActionIconToggle } from '@talend/react-components';
```
Comment (jmfrancois) [inline packages/faceted-search/src/components/FacetedSearchIcon/FacetedSearchIcon.component.js]: same here

File: packages/faceted-search/src/service/CRUDBadges.js [hunk 12]
```diff
@@ -0,0 +1,61 @@
+import cloneDeep from 'lodash/cloneDeep';
```
Comment (jmfrancois) [inline packages/faceted-search/src/service/CRUDBadges.js]: This file do not look like a service. may be just a badge.js on the package root.

File: packages/faceted-search/src/controlled.js [hunk 13]
```diff
@@ -0,0 +1,12 @@
+const isDefined = arg => typeof arg !== 'undefined';
+
+// eslint-disable-next-line import/prefer-default-export
```
[the remark this answers is not in the document]
Comment (nicomaligne) [reply packages/faceted-search/src/controlled.js]: I prefer it over default one, more opinionated.

File: packages/faceted-search/src/components/AdvancedSearch/AdvancedSearch.scss [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+
+.adv-search {
+	position: relative;
+	width: inherit;
+	height: inherit;
+
+	// Buttons basic / advanced are in absolute position,
+	// relative to the faceted toolbar.
+	// This padding is here to give them space.
+	> form {
```
[the remark this answers is not in the document]
Comment (nicomaligne) [reply packages/faceted-search/src/components/AdvancedSearch/AdvancedSearch.scss]: Yep... 😢

File: packages/faceted-search/src/components/AddFacetPopover/AddFacetPopover.component.js [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+					/>
+				))}
+			</RichLayout.Body>
+		</div>
+	);
+};
+
+AddFacetPopover.propTypes = {
+	id: PropTypes.string.isRequired,
+	initialFilterValue: PropTypes.string,
```
[the remark this answers is not in the document]
Comment (nicomaligne) [reply packages/faceted-search/src/components/AddFacetPopover/AddFacetPopover.component.js]: Nop, what do you have in mind ?

File: packages/faceted-search/src/components/context/badgeFaceted.context.js [hunk 16]
```diff
@@ -0,0 +1,17 @@
+import { createContext, useContext } from 'react';
```
[the remark this answers is not in the document]
Comment (nicomaligne) [reply packages/faceted-search/src/components/context/badgeFaceted.context.js]: Context are linked to components IMHO. But open to discuss it ;)

File: packages/faceted-search/src/components/FacetedSearch/FacetedSearch.component.js [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+	);
+
+	const isControlled = controlled('FacetedSearch', facetedMode, setFacetedMode);
+	const [facetedModeState, setFacetedModeState] = useState(FACETED_MODE.BASIC);
+	const { t } = useTranslation(I18N_DOMAIN_FACETED_SEARCH);
+	const facetedId = `${id}-faceted`;
+	const mode = isControlled ? facetedMode : facetedModeState;
+	const setMode = isControlled ? setFacetedMode : setFacetedModeState;
+	return (
+		<FacetedManager error={error} id={facetedId} inProgress={inProgress} t={t}>
```
[the remark this answers is not in the document]
Comment (nicomaligne) [reply packages/faceted-search/src/components/FacetedSearch/FacetedSearch.component.js]: Nop this is the top component of faceted, the definitions are ponly used by the sub component basic search

File: packages/faceted-search/src/components/FacetedToolbar/FacetedToolbar.component.js [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+	<div className={theme('tc-faceted-toolbar-buttons')}>
+		<Action
+			onClick={onClickAdvanced}
```
[the remark this answers is not in the document]
Comment (nicomaligne) [reply packages/faceted-search/src/components/FacetedToolbar/FacetedToolbar.component.js]: Yep sure, it has been overthink ;)

File: packages/faceted-search/src/controlled.js [hunk 19]
```diff
@@ -0,0 +1,12 @@
+const isDefined = arg => typeof arg !== 'undefined';
```
[the remark this answers is not in the document]
Comment (vbocquet) [reply packages/faceted-search/src/controlled.js]: Understand, no it's ok for me 👍

File: packages/faceted-search/src/components/types/badgeDefinition.type.js [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+	},
+	metadata: {
+		badgeId: '',
+		badgesPerFacet: 1,
+		entitiesPerBadge: 'N',
+	},
+};
+
+const fromRaw = badgeDefinitionRaw => {
+	const newStruct = Object.create(facetedBadgeDefinitionStruct);
```
[the remark this answers is not in the document]
Comment (nicomaligne) [reply packages/faceted-search/src/components/types/badgeDefinition.type.js]: I like the idea to create a proper structure, especially if we think in a typescript way. We can discuss it.

File: packages/faceted-search/src/components/AdvancedSearch/AdvancedSearch.scss (hunk 14 above)
[the remark this answers is not in the document]
Comment (vbocquet) [reply packages/faceted-search/src/components/AdvancedSearch/AdvancedSearch.scss]: 👍

File: packages/faceted-search/src/components/FacetedToolbar/FacetedToolbar.component.js (hunk 2 above)
Comment (nicomaligne) [reply packages/faceted-search/src/components/FacetedToolbar/FacetedToolbar.component.js]: Yep adv will not be convert in basic

File: packages/faceted-search/src/components/facetedSearch.propTypes.js (hunk 5 above)
Comment (nicomaligne) [reply packages/faceted-search/src/components/facetedSearch.propTypes.js]: nb of badges allowed per facet. For example you can have multiple badge name, but only one connection type.

File: packages/faceted-search/src/queryClient/index.js (hunk 9 above)
Comment (nicomaligne) [reply packages/faceted-search/src/queryClient/index.js]: Yep it's the lonk between badges and query language

File: packages/faceted-search/src/components/context/badgeFaceted.context.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (vbocquet) [reply packages/faceted-search/src/components/context/badgeFaceted.context.js]: Ok for me, it was just a suggestion based on my habits, nothing important

File: packages/faceted-search/src/service/CRUDBadges.js (hunk 12 above)
Comment (nicomaligne) [reply packages/faceted-search/src/service/CRUDBadges.js]: What do you define by service ?

File: packages/faceted-search/src/components/types/badgeDefinition.type.js (hunk 20 above)
[the remark this answers is not in the document]
Comment (vbocquet) [reply packages/faceted-search/src/components/types/badgeDefinition.type.js]: You're right.
After re-reading my comment, i don't think it make it more readable

[the remark this answers is not in the document]
Comment (vbocquet) [reply packages/faceted-search/src/components/types/badgeDefinition.type.js]: You're right.
After re-reading my comment, i don't think it make it more readable

File: packages/faceted-search/package.json [hunk 21]
```diff
@@ -5,22 +5,22 @@
...
-    "prepublish": "babel -d lib ./src/ && rimraf lib/**/*.test.js && cpx -v \"./src/**/*.scss\" ./lib",
-    "start": "start-storybook -p 6010",
-    "test": "jest --silent",
-    "test:noisy": "jest",
-    "test:watch": "jest --watch",
-    "test:cov": "jest --coverage",
+    "test": "cross-env TZ=Europe/Paris jest",
+    "test:watch": "cross-env TZ=Europe/Paris jest --watch",
+    "test:cov": "cross-env TZ=Europe/Paris jest --coverage",
+    "test:slimerjs": "slimerjs stories/slimer.js",
```
Comment (romainseb) [inline packages/faceted-search/package.json]: It's not list in the dependencies
Provide documentation about it if we want to use it

**Pull request merged**